categories.go

```diff
@@ -26,129 +28,163 @@ type Category struct {
 			notFoundError(w, err)
 			return
 		}
 
-		query := r.URL.Query()
-		var experimental bool
-		var includePolicyTemplates bool
-		var kibanaVersion *semver.Version
-
-		// Read query filter params which can affect the output
-		if len(query) > 0 {
-			if v := query.Get("experimental"); v != "" {
-				experimental, err = strconv.ParseBool(v)
-				if err != nil {
-					badRequest(w, fmt.Sprintf("invalid 'experimental' query param: '%s'", v))
-					return
-				}
-			}
+		packageList := filter.Filter(r.Context(), packages)
+		categories := filter.FilterCategories(r.Context(), packageList)
 
-			if v := query.Get("kibana.version"); v != "" {
-				kibanaVersion, err = semver.NewVersion(v)
-				if err != nil {
-					badRequest(w, fmt.Sprintf("invalid Kibana version '%s': %s", v, err))
-					return
-				}
-			}
+		data, err := getCategoriesOutput(r.Context(), categories)
+		if err != nil {
+			notFoundError(w, err)
+			return
+		}
 
-			if v := query.Get("include_policy_templates"); v != "" {
-				includePolicyTemplates, err = strconv.ParseBool(v)
-				if err != nil {
-					badRequest(w, fmt.Sprintf("invalid 'include_policy_templates' query param: '%s'", v))
-					return
-				}
-			}
+		cacheHeaders(w, cacheTime)
+		jsonHeader(w)
+		fmt.Fprint(w, string(data))
+	}
+}
+
+type categoriesFilter struct {
+	Experimental           bool
+	KibanaVersion          *semver.Version
+	IncludePolicyTemplates bool
+}
+
+func newCategoriesFilterFromParams(r *http.Request) (categoriesFilter, error) {
+	var filter categoriesFilter
+
+	query := r.URL.Query()
+	if len(query) == 0 {
+		return filter, nil
+	}
+
+	var err error
+	if v := query.Get("kibana.version"); v != "" {
+		filter.KibanaVersion, err = semver.NewVersion(v)
+		if err != nil {
+			return filter, fmt.Errorf("invalid Kibana version '%s': %w", v, err)
 		}
+	}
 
-		packageList := map[string]util.Package{}
-		// Get unique list of newest packages
-		for _, p := range packages {
-			// Check if the package is compatible with Kibana version
-			if kibanaVersion != nil {
-				if valid := p.HasKibanaVersion(kibanaVersion); !valid {
-					continue
-				}
-			}
+	if v := query.Get("experimental"); v != "" {
+		filter.Experimental, err = strconv.ParseBool(v)
+		if err != nil {
+			return filter, fmt.Errorf("invalid 'experimental' query param: '%s'", v)
+		}
+	}
 
-			// Skip internal packages
-			if p.Internal {
-				continue
-			}
+	if v := query.Get("include_policy_templates"); v != "" {
+		filter.IncludePolicyTemplates, err = strconv.ParseBool(v)
+		if err != nil {
+			return filter, fmt.Errorf("invalid 'include_policy_templates' query param: '%s'", v)
+		}
+	}
 
-			// Skip experimental packages if flag is not specified
-			if p.Release == util.ReleaseExperimental && !experimental {
-				continue
-			}
+	return filter, nil
+}
+
+func (filter categoriesFilter) Filter(ctx context.Context, packages util.Packages) map[string]util.Package {
+	span, _ := apm.StartSpan(ctx, "FilterPackages", "custom")
```

Comment:
I would suggest changing all the span types from "custom" to "app". This should work better with the "time spent by span type" chart in APM:

![image](https://user-images.githubusercontent.com/843579/127133195-e7bc8957-c40c-41db-ad63-bb58427e6672.png)

Transaction time is countered under "app", and if you add spans with type "app" then they will add to this. i.e. it's best practice (probably not written down anywhere) to use "app" for spans related to application code.

Reply:
Changed :+1: 

Btw, when initializing the service ([here](https://github.com/elastic/package-registry/pull/702/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R88)), I am using "backend.init" as transaction type, is this ok?
Are there some guidelines or list of recommended values for transaction and span types?

Reply:
That is okay, and I don't think it particularly matters. HTTP requests will be given the transaction type "request". I'm not aware of any guidance for transaction types for this kind of program initialisation.